zcash_client_backend/src/proto.rs

```diff
@@ -177,3 +196,249 @@ impl service::TreeState {
+    /// representation.
+    pub fn from_standard_proposal<P: Parameters, NoteRef>(
+        params: &P,
+        value: &Proposal<StandardFeeRule, NoteRef>,
+    ) -> Option<Self> {
+        let transaction_request = value.transaction_request().to_uri(params)?;
+        let inputs = value
+            .sapling_inputs()
+            .iter()
+            .map(|rec_note| proposal::ProposedInput {
+                txid: rec_note.txid().as_ref().to_vec(),
+                pool: 2,
+                index: rec_note.output_index().into(),
+                value: rec_note.value().into(),
+            })
+            .chain(
+                value
+                    .transparent_inputs()
+                    .iter()
+                    .map(|utxo| proposal::ProposedInput {
+                        txid: utxo.outpoint().hash().to_vec(),
+                        pool: 0,
+                        index: utxo.outpoint().n(),
+                        value: utxo.txout().value.into(),
+                    }),
+            )
+            .collect();
+
+        let balance = Some(proposal::TransactionBalance {
+            proposed_change: value
+                .balance()
+                .proposed_change()
+                .iter()
+                .map(|cv| match cv {
+                    ChangeValue::Sapling { value, memo } => proposal::ChangeValue {
+                        value: Some(proposal::change_value::Value::SaplingValue(
+                            proposal::SaplingChange {
+                                amount: (*value).into(),
+                                memo: memo.as_ref().map(|memo_bytes| proposal::MemoBytes {
+                                    value: memo_bytes.as_slice().to_vec(),
+                                }),
+                            },
+                        )),
+                    },
+                })
+                .collect(),
+            fee_required: value.balance().fee_required().into(),
+        });
+
+        #[allow(deprecated)]
+        Some(proposal::Proposal {
+            proto_version: PROPOSAL_SER_V1,
+            transaction_request,
+            inputs,
+            balance,
+            fee_rule: match value.fee_rule() {
+                StandardFeeRule::PreZip313 => proposal::FeeRule::PreZip313,
+                StandardFeeRule::Zip313 => proposal::FeeRule::Zip313,
+                StandardFeeRule::Zip317 => proposal::FeeRule::Zip317,
+            }
+            .into(),
+            min_confirmations: value.min_confirmations().into(),
+            min_target_height: value.min_target_height().into(),
+            min_anchor_height: value.min_anchor_height().into(),
+            is_shielding: value.is_shielding(),
+        })
+    }
+
+    /// Attempts to parse a [`Proposal`] based upon a supported [`StandardFeeRule`] from its
+    /// protobuf representation.
+    pub fn try_into_standard_proposal<P: consensus::Parameters, DbT: WalletRead>(
```

Comment:
It looks like this is not tested.

Reply:
That's puzzling, it should be covered in the roundtrip tests. I'll double-check.